interface/src/app/analyze/analysis/sampleBin.js

```diff
@@ -301,36 +389,50 @@ function($log, $cacheFactory, $q, Sample, Activity, NodeInfo, MathFuncts) {
-          //      `activityCache` [inner .map()]
-          // FIXME: counting on array order to match node order here
-          return this.activityCache.get(sampId)[val.node - 1].value;
-        };
-        var nodeObject = {
-          'id': val.node, // TODO: map this id to node name (via vega spec?)
-          'activityA': sg['group-a'].map(mapSampleIdsToActivity, this),
-          'activityB': sg['group-b'].map(mapSampleIdsToActivity, this)
-        };
-        nodeObject.diff = (
-          MathFuncts.mean(nodeObject.activityB) -
-          MathFuncts.mean(nodeObject.activityA)
-        );
-        nodeObject.logsig = -Math.log10(MathFuncts.tTest(
-          nodeObject.activityA, nodeObject.activityB
-        ).pValue());
+      // (1a) we obtain a list of nodes by retrieving node activity
+      //      for the first sample in our volcano plot
+      var firstSampleNodes = this.activityCache.get(sg['group-a'][0]).map(
+        function(val) {
+          return val.node;  // extract just the node IDs
+        }
+      );
+      // (1b) now obtain (and cache) a name for each node id
+      var nodeInfoSetPromise = this.getNodeInfoSetPromise(firstSampleNodes);
+      var mapNodesToNodeInfo = function() {
+        // (2a) next, we build an array (replacing `volcanoData.source`)
+        //      comprised of `nodeObject`s by walking through the
+        //      `firstSampleNodes` and constructing a `nodeObject` for
+        //      each. [outer .map()]
+        cbSampleBin.volcanoData.source = firstSampleNodes.map(function(id) {
+          var mapSampleIdsToActivity = function(sampId) {
+            // (2b) the array of activity for each node is built by plucking the
+            //      activity `.value` for each sample within this node from the
+            //      `activityCache` [inner .map()]
+            // FIXME: counting on array order to match node order here
+            return cbSampleBin.activityCache.get(sampId)[id - 1].value;
+          };
+          var nodeObject = {
+            'id': id,
+            'name': cbSampleBin.getCachedNodeInfo(id).name,
+            'activityA': sg['group-a'].map(mapSampleIdsToActivity),
+            'activityB': sg['group-b'].map(mapSampleIdsToActivity)
+          };
+          nodeObject.diff = (
+            MathFuncts.mean(nodeObject.activityA) -
+            MathFuncts.mean(nodeObject.activityB)
+          );
+          nodeObject.logsig = -Math.log10(MathFuncts.tTest(
+            nodeObject.activityA, nodeObject.activityB
+          ).pValue());
 
-        return nodeObject;
-      }, this);
-      return retval;
+          return nodeObject;
+        });
+        // no return needed here: we've updated `cbSampleBin.volcanoData`
+      };
+      // invoke mapNodesToNodeInfo only after nodeInfoSetPromise is fulfilled
+      nodeInfoSetPromise
+        .then(mapNodesToNodeInfo)
+        .catch(this.logError);
     }
   };
```

Comment:
I first defined `sampleBin.js` to be an object for adding, removing and sorting the samples that are managed by the "sample bin" widget. Then I built the heatmap and when we needed to do clustering I had performance issues so I added caching and clustering to `sampleBin`. I think you can see where this is going.

As I built the volcano plot, I hoped to do so in a more "clean" way to get away from overloading `sampleBin` so badly, but I couldn't pull that off properly without refactoring to separate caching from `sampleBin`.

I now see how to do this (node and activity services should be broken out and have smart caching built in, sampleBin should return to its core function and heatmap and volcano plot features belong in their own services as well), but I have to finish off the volcano plot first (this PR and one more to get the selected nodes back out of vega) and add some unit tests to be sure my refactoring doesn't break lots of stuff. Issue #76.